File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys
n, m = map(int, sys.stdin.readline().split())
tree = list(map(int, sys.stdin.readline().split()))

start, end = 1, max(tree)

while True:
    # print(start, end)
    if start > end:
        break
    # 목재절단기의 높이
    mid = (start + end) // 2
    pieces = 0
    for i in range(n):
        if tree[i] - mid > 0:
            pieces += tree[i] - mid
    # print(pieces)
    # 조각이 너무 많이 나와서 mid를 늘려야함
    if pieces >= m:
        start = mid + 1
    else:
        end = mid - 1
print(end)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int[] tree = new int[n];
        int maxtree = 0;
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            tree[i] = Integer.parseInt(st.nextToken());
            maxtree = Math.max(tree[i], maxtree);
        }

        int start = 0;
        int end = maxtree;

        while (start <= end) {
            int mid = (start + end) / 2;
            long take = 0;
            for (int i = 0; i < n; i++) {
                if (tree[i] > mid) {
                    take += tree[i] - mid;
                }
            }
            if (take >= m) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        System.out.println(end);
    }


}